pkg/cli/cachedaemon.go

```diff
@@ -89,43 +90,68 @@ func NewCacheDaemonCommand() *cobra.Command {
 				shutdownTelemetry = telemetry.Init(ctx, telemetry.Server)
 			}
 
-			cert, err := tls.LoadX509KeyPair(certFile, keyFile)
-			if err != nil {
-				return fmt.Errorf("cannot open TLS cert and key files (%s, %s): %w", certFile, keyFile, err)
-			}
-
-			pasetoKey, err := parsePublicKey(pasetoFile)
-			if err != nil {
-				return fmt.Errorf("cannot parse Paseto public key %s: %w", pasetoFile, err)
-			}
-
 			cl, err := client.NewClient(ctx, upstreamHost, upstreamPort, client.WithheadlessHost(headlessHost))
 			if err != nil {
 				return err
 			}
 			ctx = client.IntoContext(ctx, cl)
 
-			listen, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
-			if err != nil {
-				return fmt.Errorf("failed to listen on TCP port %d: %w", port, err)
-			}
+			var s *cached.CacheServer
+			if csiSocket == "" {
+				cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+				if err != nil {
+					return fmt.Errorf("cannot open TLS cert and key files (%s, %s): %w", certFile, keyFile, err)
+				}
+				pasetoKey, err := parsePublicKey(pasetoFile)
+				if err != nil {
+					return fmt.Errorf("cannot parse Paseto public key %s: %w", pasetoFile, err)
+				}
 
-			s := cached.NewServer(ctx, cl, &cert, stagingPath, pasetoKey)
+				s = cached.NewServer(ctx, cl, &cert, stagingPath, pasetoKey)
+			} else {
+				s = cached.NewCacheCSIServer(ctx, cl, stagingPath)
+			}
 
 			osSignals := make(chan os.Signal, 1)
+			errors := make(chan error, 2)
 			signal.Notify(osSignals, os.Interrupt, syscall.SIGTERM)
-			go func() {
-				<-osSignals
-				s.Grpc.GracefulStop()
-			}()
 
 			err = s.Cached.Prepare(ctx)
 			if err != nil {
 				return fmt.Errorf("failed to prepare cache daemon in %s: %w", stagingPath, err)
 			}
 
-			logger.Info(ctx, "start server", key.Port.Field(port), key.Environment.Field(env.String()))
-			return s.Serve(listen)
+			if csiSocket != "" {
+				go (func() {
+					logger.Info(ctx, "start CSI server")
+					err := s.ServeCSI(ctx, csiSocket)
```

Comment:
I don't think this will cleanly stop on graceful shutdown. You likely want a cancellable context passed in here that we can abort on `<-osSignals`

Reply:
Would you mind terribly shooting me a diff? I feel like I could figure it out but it would take you approximately 10 seconds to do